x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/SecondaryAuthentication.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.core.security.authc.support;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+import org.elasticsearch.rest.RestRequest;
+import org.elasticsearch.transport.TransportRequest;
+import org.elasticsearch.xpack.core.security.SecurityContext;
+import org.elasticsearch.xpack.core.security.authc.Authentication;
+
+import java.io.IOException;
+import java.util.Objects;
+import java.util.function.Function;
+
+/**
+ * Some Elasticsearch APIs need to be provided with 2 sets of credentials.
+ * Typically this happens when a system user needs to perform an action while accessing data on behalf of, or user information regarding
+ * a logged in user.
+ * This class is a representation of that secondary user that can be activated in the security context while processing specific blocks
+ * of code or within a listener.
+ */
+public class SecondaryAuthentication {
+
+    private static final String THREAD_CTX_KEY = "_xpack_security_2nd_authc";
+
+    public interface Authenticator {
```

Comment:
It existed so that x-pack-core had something usable even though the implementation was in security.
But actually that doesn't matter anymore because the only use the authentication from within security.